add unique index for vcategory owncloud/core#20952

cortho opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
add unique key for table *dbprefix*vcategory

## Related Issue
https://github.com/owncloud/core/issues/20952

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
fix as proposed at https://github.com/owncloud/core/issues/20952#issuecomment-169949310

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

cortho committed:
add unique index for vcategory owncloud/core#20952

cortho committed:
add unique index for vcategory owncloud/core#20952

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/core?pullRequest=27913) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/core?pullRequest=27913) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/core?pullRequest=27913) it.</sub>

DeepDiver1975 commented:
THX - but I see following issues:
- [ ] applying the index will fail on upgrade where data is already in the system
- [ ] please use migrations to change the schema - https://doc.owncloud.org/server/10.0/developer_manual/app/schema.html#database-migrations

DeepDiver1975 reviewed (changes requested):
see the comment ....

cortho committed:
add migration owncloud/core#20952

cortho committed:
add migration owncloud/core#20952

cortho commented:
added migration in https://github.com/cortho/core/commit/d674ff0169dd23164c6c4f190ff3c1e5da039993
what to do with db_structure.xml I changed in the previous commit?

cortho committed:
Merge branch 'vcategory_unique_index' of github.com:cortho/core into vcategory_unique_index

cortho committed:
fix jenkins ORA-00972 for owncloud/core#20952

DeepDiver1975 commented on core/Migrations/Version20170519112910.php:
```diff
@@ -0,0 +1,69 @@
+            $qb = $connection->getQueryBuilder();
+            $qb->update('vcategory_to_object')
+            ->set('categoryid', $qb->createNamedParameter($key))
+            ->where($qb->expr()->in(
```
will break in case the array has more then 1000 elements - dbms have restrictions on this .....

DeepDiver1975 commented on core/Migrations/Version20170519112910.php:
```diff
@@ -0,0 +1,69 @@
+        $duplicates = call_user_func_array('array_merge', $duplicates);
+        $qb = $connection->getQueryBuilder();
+        $qb->delete('vcategory')
+        ->where($qb->expr()->in(
```
same as above

DeepDiver1975 commented on db_structure.xml:
```diff
@@ -966,6 +966,23 @@
 				</field>
 			</index>
 
+			<index>
```
you can remove this change

cortho replied:
Thank you very much for your feedback, I'll have a look on this.

cortho committed:
chunk arrays owncloud/core#20952

PVince81 commented:
@DeepDiver1975 PR was updated, please recheck
